pkg/controller/applicationlayer/applicationlayer_controller.go

```diff
@@ -0,0 +1,249 @@
+	// create the reconciler
+	reconciler := newReconciler(mgr, opts, licenseAPIReady)
+
+	// Create a new controller
+	c, err := controller.New("applicationlayer-controller", mgr, controller.Options{Reconciler: reconcile.Reconciler(reconciler)})
+	if err != nil {
+		return err
+	}
+
+	k8sClient, err := kubernetes.NewForConfig(mgr.GetConfig())
+	if err != nil {
+		log.Error(err, "Failed to establish a connection to k8s")
+		return err
+	}
+
+	go utils.WaitToAddLicenseKeyWatch(c, k8sClient, log, licenseAPIReady)
+
+	return add(mgr, c)
+}
+
+// newReconciler returns a new *reconcile.Reconciler
+func newReconciler(mgr manager.Manager, opts options.AddOptions, licenseAPIReady *utils.ReadyFlag) reconcile.Reconciler {
+	r := &ReconcileApplicationLayer{
+		client:          mgr.GetClient(),
+		scheme:          mgr.GetScheme(),
+		provider:        opts.DetectedProvider,
+		status:          status.New(mgr.GetClient(), "applicationlayer", opts.KubernetesVersion),
+		clusterDomain:   opts.ClusterDomain,
+		licenseAPIReady: licenseAPIReady,
+	}
+	r.status.Run()
+	return r
+}
+
+// add adds watches for resources that are available at startup
+func add(mgr manager.Manager, c controller.Controller) error {
+	var err error
+
+	// Watch for changes to primary resource applicationlayer
+	err = c.Watch(&source.Kind{Type: &operatorv1.ApplicationLayer{}}, &handler.EnqueueRequestForObject{})
+	if err != nil {
+		return err
+	}
+
+	if err = utils.AddNetworkWatch(c); err != nil {
+		return fmt.Errorf("applicationlayer-controller failed to watch Network resource: %w", err)
+	}
+
+	if err = imageset.AddImageSetWatch(c); err != nil {
+		return fmt.Errorf("applicationlayer-controller failed to watch ImageSet: %w", err)
+	}
+
+	if err = utils.AddAPIServerWatch(c); err != nil {
+		return fmt.Errorf("applicationlayer-controller failed to watch APIServer resource: %w", err)
+	}
+	return nil
+}
+
+// blank assignment to verify that ReconcileCompliance implements reconcile.Reconciler
+var _ reconcile.Reconciler = &ReconcileApplicationLayer{}
+
+// ReconcileApplicationLayer reconciles a ApplicationLayer object
+type ReconcileApplicationLayer struct {
+	// This client, initialized using mgr.Client() above, is a split client
+	// that reads objects from the cache and writes to the apiserver
+	client          client.Client
+	scheme          *runtime.Scheme
+	provider        operatorv1.Provider
+	status          status.StatusManager
+	clusterDomain   string
+	licenseAPIReady *utils.ReadyFlag
+}
+
+// Reconcile reads that state of the cluster for a ApplicationLayer object and makes changes
+// based on the state read and what is in the ApplicationLayer.Spec
+func (r *ReconcileApplicationLayer) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	reqLogger := log.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
+	reqLogger.Info("Reconciling ApplicationLayer")
+
+	instance, err := GetApplicationLayer(ctx, r.client)
```

Comment:
We need to include a defaulting step below for this resource. You can see in the core or authentication controllers how we do that.